File: JavaAlgorithm/src/main/java/algo/tzashinorpu/Daily/COM/detectCycle_142.java
package algo.tzashinorpu.Daily.COM;

public class detectCycle_142 {
    public ListNode detectCycle(ListNode head) {
        ListNode ptr = head;
        ListNode slow = head;
        ListNode fast = head;
        while (slow != null && fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            // slow 和 fast 相遇后  slow 到入环的距离 = head 到入环的距离
            // https://leetcode.com/problems/linked-list-cycle-ii/discuss/44774/Java-O(1)-space-solution-with-detailed-explanation.
            if (slow == fast) {
               /* if (ptr == slow) {
                    return ptr;
                } else {
                    ptr = ptr.next;
                    slow = slow.next;
                }*/
                while (ptr != slow) {
                    ptr = ptr.next;
                    slow = slow.next;
                }
                return ptr;
            }
        }
        return null;
    }
}

File: Clion/labuladong/chapter0/linklist/detectCycle_142.cpp
//
// Created by TzashiNorpu on 5/31/2023.
//
#include "../../header/header.h"

using namespace labuladong;

ListNode *Solution::detectCycle(ListNode *head) {
  ListNode *slow = head;
  ListNode *fast = head;
  while (fast && fast->next) {
    slow = slow->next;
    fast = fast->next->next;
    if (fast == slow) {
      ListNode *temp = head;
      while (temp != slow) {
        temp = temp->next;
        slow = slow->next;
      }
      return temp;
    }
  }
  return nullptr;
}